edn/src/types_pp.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![cfg_attr(feature = "cargo-clippy", allow(linkedlist))]
+
+use std::collections::{BTreeSet, BTreeMap, LinkedList};
+use std::borrow::Cow;
+use pp::{PrettyPrintTree, PrintInstr};
+use types::Value;
+
+// A pretty printing implementation for an `edn::Value`, focused on limited
+// whitespace expansion and readability for mentat queries.
+impl<'a> From<&'a Value> for PrettyPrintTree<'a> {
+    fn from(src: &Value) -> PrettyPrintTree {
+        match *src {
+            Value::Vector(ref v) => v.into(),
+            Value::List(ref v) => v.into(),
+            Value::Set(ref v) => v.into(),
+            Value::Map(ref v) => v.into(),
+            Value::Text(ref v) => string_leaf!(format!("\"{}\"", v)),
+            _ => string_leaf!(src.to_string())
+        }
+    }
+}
+
+impl<'a> From<&'a Vec<Value>> for PrettyPrintTree<'a> {
```

Comment:
This seems a pretty big abstraction leak.  We want to be able to print `edn::Value`, not arbitrary vectors (or maps, or sets) of values.  I see you're using `From` throughout, but I think you should use it only for the top-level object.